### DISCUSSION TITLE
ypy-in-pyodide Quill demo

### DISCUSSION THREAD

**Discussion by kafonek** [y-crdt/ypy#107] [Show and tell]
I took a shot at recreating the [yjs Quill example](https://demos.yjs.dev/quill/quill.html) using y-py in a webworker with pyodide. Source code at https://github.com/kafonek/quill-demo. I recorded a short demo to share internally at Noteable, so some of the comments in the video are for that audience but here's yjs, server-side ypy, and pyodide-ypy in action - https://www.youtube.com/watch?v=eOlydiiWX0E

Some thoughts about the experience.
 - Quill is neat, the delta structure is cool. I had been frustrated by not having a minimal "yjs bindings to a textarea" example but now I totally understand why minimal examples with frameworks like Quill are better
 - I was happy with how much control I had in Python, from installing dependencies to using "native js websocket" with Python syntax to registering callbacks for messages coming in on websocket or webworker
 - I wish I had access to the quill canvas object in my webworker. That would cut down on the complexity of writing message handlers on both sides of the webworker (`index.html` and `worker.py`)
 - I see the point of `origin` argument in transactions now! Need to get that into `yrs` / `y-py`
 - I think it makes sense to put the message serialization helper code (`ypy-websockets.yutils`) into `yrs`

There's lots of room to build on this current demo, and I want to explore `codemirror` bindings next since we use that framework. Hope everything works if you try it on your machine, let me know if you run into any bumps! Thanks.

**Comment by davidbrochart** [comment]
Quite impressive!
I'm wondering if your approach could be used in https://github.com/jupyterlite/jupyterlite/issues/896. But I think we don't want to pay the cost of a Python interpreter just to enable collaboration.

> * I think it makes sense to put the message serialization helper code (`ypy-websockets.yutils`) into `yrs`

You mean into `ypy`? Yes, I agree, since it's network-agnostic.

> I want to explore `codemirror` bindings next since we use that framework

You should look at how it's done in JupyterLab.

**Reply by kafonek** [reply]
Thanks @davidbrochart. I actually wanted to tease this little demo at the Jupyterlab realtime community meeting this morning but I think that's another out of date/dead calendar item :(  I'll read up on jupyterlab code. It'd be great to pair with someone experienced in that code base sometime. I agree that pyodide is too much of a cost right now. I hope things will change in the future.

I suggested putting the `yutils` helper methods in `yrs` so that they wouldn't have to be implemented in each language binding (`ypy`, ruby, etc) and I figure it would be more performant in Rust. That being said, I don't have enough experience to write it up in Rust and moving from `ypy-websockets` to `y-py` alone would be an incremental improvement imo.
